### PR TITLE
feat: OSS and IaC delta matcher

### PR DESCRIPTION
**Pull request by ShawkyZ**
### Description

* Move files under domain/snyk package to subpackages.
* Refactor base branch checking out logic.
* Move base branch scanning to scanners.go.
* Enable Code, OSS, IaC scanners to be delta matchers.
* Enrich fingerprints for OSS and IaC.
* Rename code_matcher to fuzzy_matcher since we can use it for other products.

With this change OSS and IaC should send back deltas if flag is enabled.

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [ ] README.md updated, if user-facing
- [ ] License file updated, if new 3rd-party dependency is introduced

### CODE REVIEW

File: domain/snyk/issues.go [hunk 1]
```diff
@@ -75,10 +75,19 @@ type Issue struct {
...
+		return nil
+	}
+	return ossIssueData.From
+}
+
 var _ delta.Identifiable = (*Issue)(nil)
 var _ delta.Fingerprintable = (*Issue)(nil)
 var _ delta.Locatable = (*Issue)(nil)
 var _ delta.Pathable = (*Issue)(nil)
+var _ delta.Fromable = (*Issue)(nil)
```
Comment (bastiandoetsch) [inline domain/snyk/issues.go:90]: I'd like something more specific, that does not imply knowledge of the json output and its field names - something like "introducedBy" maybe?

File: domain/snyk/issues.go [hunk 2]
```diff
@@ -363,18 +273,21 @@ func (i *Issue) UnmarshalJSON(data []byte) error {
 			return err
 		}
 		i.AdditionalData = codeData
+		break
```
Comment (bastiandoetsch) [inline domain/snyk/issues.go:276]: Why? In Go, afaik, you never need a break after a case - the default is break, you need to specify `fallthrough` if you need a fallthrough.

File: internal/delta/code_matcher.go [hunk 3]
```diff
@@ -45,6 +49,7 @@ var weights = struct {
 	FilePositionDistance        float64
 	RecentHistoryDistance       float64
 	FingerprintConfidence       float64
+	FromListConfidence          float64
```
Comment (bastiandoetsch) [inline internal/delta/code_matcher.go:52]: Same here, I'd name it `dependencyChain` or `supplyChain` confidence.

File: internal/delta/findings_identifiable.go [hunk 4]
```diff
@@ -21,6 +21,11 @@ type Fingerprintable interface {
 	GetFingerprint() string
 }
 
+type Fromable interface {
```
Comment (bastiandoetsch) [inline internal/delta/findings_identifiable.go:24]: see above

File: infrastructure/oss/cli_scanner.go [hunk 5]
```diff
@@ -169,7 +173,7 @@ func (cliScanner *CLIScanner) scanInternal(ctx context.Context, path string, com
 	defer p.EndWithMessage("Snyk Open Source scan completed.")
 
 	path, err := filepath.Abs(path)
-	if errorInfo != nil {
+	if err != nil {
```
Comment (ShawkyZ) [inline infrastructure/oss/cli_scanner.go:176]: this was never nil. that's why I made the change here.

File: domain/snyk/scanner/scanner.go [hunk 6]
```diff
@@ -287,8 +292,100 @@ func (sc *DelegatingConcurrentScanner) Scan(
...
+			return foundIssues, nil // Returning an empty slice implies that no issues were found
+		}
+	}
+
+	foundIssues, err = s.Scan(ctx, path, folderPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if enabled, _ := isDeltaScanEnabled(s); enabled && len(foundIssues) > 0 {
```
Comment (bastiandoetsch) [inline domain/snyk/scanner/scanner.go:337]: please re-use check result from above :)

**Review by bastiandoetsch — APPROVED** [review]

**Pull request merged**